client/python/cache_map.py

```diff
@@ -0,0 +1,184 @@
+# Copyright 2014 Treode, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+from llist import *
+from tx_clock import *
+
+class _InternalCacheObject(object):
+    def __eq__(self, other):
+        if (isinstance(other, _InternalCacheObject)):
+            return self.__dict__ == other.__dict__
+        return False
+
+    def __hash__(self):
+        return hash(repr(self))
+
+    def __repr__(self):
+        return type(self).__name__ + str(self.__dict__)
+
+"""
+object returned by get method; type of cache_map values
+"""
+class CacheResult(_InternalCacheObject):
+    def __init__(self, value_time, cached_time, value):
+        self.value_time = value_time
+        self.cached_time = cached_time
+        self.value = value
+
+"""
+dictionary of cache entries implementing LRU cache
+"""
+class CacheMap(object):
+
+    class _CacheEntry(_InternalCacheObject):
+        def __init__(self, table_name, key, value, 
+            cached_time, value_time):
+            self.table_name = table_name
+            self.key = key
+            self.value = value
+            self.cached_time = cached_time
+            self.value_time = value_time
+
+    def __init__(self, max_size):
+        self.max_size = max_size
+
+        # Store values as a dictionary of lists:
+        # { (table, key): [ dllistnode(_CacheEntry), ... ]}
+        self.entry_table = {}
+
+        # Track the order of entries added to / accessed in 
+        # cache to maintain LRU policy [ dllistnode(_CacheEntry) ... ]
+        self.lru_list = dllist()
+
+    def _lru_add(self, entry):
+        node = self.lru_list.appendleft(entry)
+        self._lru_evict()
+        return node
+
+    def _lru_refresh(self, node):
+        lru_list = self.lru_list
+        entry = lru_list.remove(node)
+        new_node = lru_list.appendleft(entry)
+        self.lru_list = lru_list
+        return new_node
+
+    def _lru_evict(self):
+        entry_table = self.entry_table
+        lru_list = self.lru_list
+        max_size = self.max_size
+        remove_index = len(lru_list) - max(0, len(lru_list)-max_size)
+        for i in xrange(remove_index, len(lru_list)):
+            # Remove evicted entries from the entry_table
+            node = lru_list.nodeat(i)
```

Comment:
`nodeat` is O(n). It looks like the left is the most recent, so to evict you want to `while (size > maxsize) popright`.